controllers/imapmon.go

```diff
@@ -0,0 +1,184 @@
+//func NewImapMonitor(configP config.PhishServer, configA *config.Config) *ImapMonitor {
+func NewImapMonitor(config *config.Config) *ImapMonitor {
+	// Make sure database connection exists. Not sure why I have to do this here, but
+	//  otherwise db is <nil> when calling models.GetEnabledIMAPs()
+	err := models.Setup(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	reportURL := "http://" + config.PhishConf.ListenURL
+	if config.PhishConf.UseTLS {
+		reportURL = "https://" + config.PhishConf.ListenURL
+	}
+
+	im := &ImapMonitor{
+		reportURL: reportURL,
+		checkFreq: config.IMAPFreq,
+	}
+	return im
+}
+
+// Start launches the IMAP campaign monitor
+func (im *ImapMonitor) Start() error {
+	log.Info("Starting IMAP monitor")
+	ctx, cancel := context.WithCancel(context.Background()) //ctx is the derivedContext
+	im.cancel = cancel
+	go PeriodicIMAPMonitor(ctx, im)
+	return nil
+}
+
+// Shutdown attempts to gracefully shutdown the IMAP monitor.
+func (im *ImapMonitor) Shutdown() error {
+	log.Info("Shutting down IMAP monitor")
+	im.cancel()
+	return nil
+}
+
+// checkForNewEmails logs into an IMAP account and checks unread emails
+//  for the rid campaign identifier.
+func checkForNewEmails(im models.IMAP, wg *sync.WaitGroup, reportURL string) {
+	defer wg.Done()
+	mailSettings := eazye.MailboxInfo{
+		Host:   im.Host,
+		TLS:    im.TLS,
+		User:   im.Username,
+		Pwd:    im.Password,
+		Folder: im.Folder}
+
+	//Make sure server ends with a trailing slash
+	if reportURL[len(reportURL)-1:] != "/" {
+		reportURL = reportURL + "/"
+	}
+	//Set http timeout to 10 seconds, or routine may hang
+	var netClient = &http.Client{
+		Timeout: time.Second * 10,
+	}
+
+	msgs, err := eazye.GetUnread(mailSettings, true, false)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	//Update last_succesful_login here via im.Host
+	err = models.SuccessfulLogin(&im)
+
+	if len(msgs) > 0 {
+		var reportingFailed []uint32 // UIDs of emails that were unable to be reported to phishing server, mark as unread
+		var campaignEmails []uint32  //UIDs of campaign emails. If deleteCampaignEmails is true, we will delete these
+		for _, m := range msgs {
+			//Check if sender is from company's domain, if enabled. TODO: Make this an IMAP filter
+			if im.RestrictDomain != "" { //e.g domainResitct = widgets.com
+				splitEmail := strings.Split(m.From.Address, "@")
+				senderDomain := splitEmail[len(splitEmail)-1]
+				if senderDomain != im.RestrictDomain {
+					log.Debug("Ignoring email as not from company domain: ", senderDomain)
+					continue
+				}
+			}
+
+			body := string(append(m.Text, m.HTML...)) //Empty body is being returned for emails forwarded from ProtonMail. Need to investigate more. Perhaps it's being attached, and eazye doesn't support attachments.
+			rid := goPhishRegex.FindString(body)
+
+			if rid != "" {
+				rid := strings.TrimSpace(rid)
+				reportURL := fmt.Sprintf("%sreport%s", reportURL, rid)
+				response, err := netClient.Get(reportURL)
```

Comment:
This is an interesting approach. On one hand, it seems odd that we're making (almost) an external HTTP request to Gophish for the reporting endpoint from within Gophish. Instead, it seems preferred that we would load the `models.Result` object and handle the email report like we do in the [report handler](https://github.com/gophish/gophish/blob/master/controllers/phish.go#L188).

But, on the other hand, the benefit this approach provides is that for logging it can appear as though someone hit the reporting endpoint. 🤷‍♂ 

Just curious to hear your thoughts here while I think a little more about the approach I'd prefer.

Reply:
I went through the same thought process. It seems neater (and faster) to update the database directly, but as you guessed I did it this way for debugging to the console.

I'm happy to update to rather call `HandleEmailReport()` if you'd prefer.